### PR TITLE
Assignment2 @ Tony Chan

### PR DESCRIPTION
**Pull request by tnadalini**
initial PR (assignment not yet completed)

### CODE REVIEW

File: Logger/FileLogger.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+                _FilePath = value;
+            }
+        }
+
+        public FileLogger(string? filePath) { FilePath = filePath; }  
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+            string logResult = $"{DateTime.Now} {ClassName} {logLevel}: {message} \n";
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:25]: String interpolation! The only suggestion I have here is to use `Environment.NewLine` instead of `\n`, but looks good otherwise. 
```suggestion
            string logResult = $"{DateTime.Now} {ClassName} {logLevel}: {message}{Environment.NewLine}";
```

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -1,11 +1,44 @@
...
-            return null;
+            get => _FilePath!;
+
+            set
+            {
+                if (value == null) { throw new ArgumentNullException(nameof(value)); }
+
+                _FilePath = value;
+            }
+        }
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:18]: Same comments I made for the `FileLogger.cs` apply here. (However, this can be nullable)

File: Logger/LogFactory.cs [hunk 3]
```diff
@@ -1,11 +1,44 @@
...
+                    ClassName = className
+                }; 
+
+                return fileLogger;
+            }
+
+            return null; 
+        }
+
+        public void ConfigureFileLogger (string filePath)
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:38]: 👍🏻

File: Logger/LogFactory.cs [hunk 4]
```diff
@@ -1,11 +1,44 @@
...
+            set
+            {
+                if (value == null) { throw new ArgumentNullException(nameof(value)); }
+
+                _FilePath = value;
+            }
+        }
+
+
+        public BaseLogger? CreateLogger(string className, string? filePath)
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:21]:
```suggestion
        public BaseLogger? CreateLogger(string className)
```

You should be using the filePath set from the `ConfigureFileLogger` method, not passing in a new filePath

File: Logger/LogFactory.cs [hunk 5]
```diff
@@ -1,11 +1,44 @@
...
+        {
+            if (filePath != null)
+            {
+                ConfigureFileLogger(filePath);
+                
+                FileLogger fileLogger = new FileLogger(FilePath)
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:27]: Notice here, you switch between using `filePath` and `FilePath`, even though you haven't verified that `FilePath` is not null. This can very easily lead to difficult-to-find bugs in your code.

File: Logger/BaseLoggerMixins.cs [hunk 6]
```diff
@@ -1,7 +1,56 @@
-﻿namespace Logger
+﻿using System;
+
+namespace Logger
 {
     public static class BaseLoggerMixins
     {
-        
+        private static string? messageWithArgs;
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:7]:
```suggestion

```

If you're going to use this as an intermediate variable between formatting and logging, you should make the scope of this variable local to the current function being called. 

Leaving it here, as a variable accessible to the scope of this entire class makes your intent less clear (confusing code) and leaves you open to possible bugs where a leftover string can be logged out instead of the string you intend.

File: Logger/BaseLoggerMixins.cs [hunk 7]
```diff
@@ -1,7 +1,56 @@
...
+        public static void Error(this BaseLogger? logger, string message, params object[] args) 
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            
+            messageWithArgs = String.Format(message, args);
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:15]:
```suggestion
            var messageWithArgs = string.Format(message, args);
```

File: Logger/BaseLoggerMixins.cs [hunk 8]
```diff
@@ -1,7 +1,56 @@
...
+        }
+
+        public static void Warning(this BaseLogger? logger, string message, params object[] args)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            messageWithArgs = String.Format(message, args);
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:27]:
```suggestion
            var messageWithArgs = string.Format(message, args);
```

File: Logger/BaseLoggerMixins.cs [hunk 9]
```diff
@@ -1,7 +1,56 @@
...
+        }
+
+        public static void Information(this BaseLogger? logger, string message, params object[] args)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            messageWithArgs = String.Format(message, args);
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:39]:
```suggestion
            var messageWithArgs = string.Format(message, args);
```

File: Logger/BaseLoggerMixins.cs [hunk 10]
```diff
@@ -1,7 +1,56 @@
...
+        }
+
+        public static void Debug(this BaseLogger? logger, string message, params object[] args)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            messageWithArgs = String.Format(message, args);
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:51]:
```suggestion
           var messageWithArgs = string.Format(message, args);
```

File: Logger/Logger.csproj [hunk 11]
```diff
@@ -2,6 +2,12 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
+	  <Nullable>enable</Nullable>
+	  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+	  <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+	  <EnableNETAnalyzers>true</EnableNETAnalyzers>
+	  <LangVersion>10.0</LangVersion>
```
Comment (mmwoodfo) [inline Logger/Logger.csproj:9]: Good job here, although you are supposed to enable these for both projects (Logger.Tests.csproj)

File: Logger.Tests/LogFactoryTests.cs [hunk 12]
```diff
@@ -1,11 +1,42 @@
...
 using System.IO;
+using System;
 
 namespace Logger.Tests
 {
     [TestClass]
     public class LogFactoryTests
     {
+        LogFactory exLogFactory; //'example LogFactory
+        FileLogger exLogger; //'example Logger'
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:11]: Again, keep these within the scopes of the individual tests
```suggestion

```

File: Logger.Tests/LogFactoryTests.cs [hunk 13]
```diff
@@ -1,11 +1,42 @@
...
+
+        [TestInitialize]
+        public void Initialize()
+        {
+            exLogFactory = new LogFactory();
+            exLogger = (FileLogger) exLogFactory.CreateLogger("fileLog", "C://Egg");
+        }
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:18]: There shouldn't be any need for a test initializer, as you probably want to make a fresh class for each test in this case. Also, you should be using `nameof()` here to get the classname.

File: Logger.Tests/LogFactoryTests.cs [hunk 14]
```diff
@@ -1,11 +1,42 @@
...
+        {
+            FileLogger logger = (FileLogger) exLogFactory.CreateLogger("fileLog1", "C://Egg");
+
+            //Assert.Equals(exLogger.ClassName, logger.ClassName);
+            Assert.AreEqual(exLogger.FilePath, logger.FilePath);
+            Assert.AreEqual("fileLog1", logger.ClassName);
+        }
+
+        [TestMethod]
+        public void NullFilePath_ThrowNullException()
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:31]:
```suggestion
        public void NullFilePath_ReturnsNull()
```

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 15]
```diff
@@ -14,19 +14,60 @@ public void Error_WithNullLogger_ThrowsException()
...
+            // Act
+            BaseLoggerMixins.Warning(null, "");
+
+            // Assert
+            //Assert.ThrowsException<ArgumentNullException>();
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void Debug_WithNullLogger_ThrowsException()
```
Comment (mmwoodfo) [inline Logger.Tests/BaseLoggerMixinsTests.cs:51]: In addition to testing the null inputs, you should also test that it logs the message correctly when the input is not null.

File: Logger.Tests/FileLoggerTests.cs [hunk 16]
```diff
@@ -5,6 +5,30 @@ namespace Logger.Tests
     [TestClass]
     public class FileLoggerTests
     {
-        
+
+        LogFactory logFactory;
+        FileLogger fileLogger;
+
+        [TestInitialize]
+        public void TestInitialize()
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:13]: You should have some tests in here that confirms the logger is appending to the end of a file, logging correct log levels, formatting correctly, etc.
